### PR TITLE
[feat] Add LinodeFirewall controller

### PR DESCRIPTION
**Pull request by AshleyDumaine**
**What this PR does / why we need it**: It should be possible to create a Cloud Firewall rather than just relying on cilium hostNetwork policies for a cluster. This adds the ability to create Cloud Firewalls but does not yet tie them to Machines or Clusters

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Example Firewall:
```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: LinodeFirewall
metadata:
  name: test-fw
spec:
  enabled: true
  inboundPolicy: DROP
  inboundRules:
    - action: ACCEPT
      label: ssh
      ports: "22"
      protocol: "TCP"
      addresses:
        ipv4:
          - "192.168.1.1/32"
```
The inboundRules and outboundRules follow https://pkg.go.dev/github.com/linode/linodego#FirewallRule 

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [x] adds or updates e2e tests

### CODE REVIEW

File: .github/workflows/e2e-test.yaml [hunk 1]
```diff
@@ -29,6 +29,7 @@ on:
           - linodeobj
           - linodevpc
           - linodeplacementgroup
+          - linodefirewall
```
Comment (komer3) [inline .github/workflows/e2e-test.yaml:32]: Not sure if you added this test to `.github/filters.yml`. We need to add it there to trigger this e2e test!

File: .golangci.yml [hunk 2]
```diff
@@ -131,7 +131,7 @@ linters:
     - decorder
     # - depguard
     - dogsled
-    - dupl
+    # - dupl
```
Comment (komer3) [inline .golangci.yml:134]: question: any reason you commented this out?

File: api/v1alpha2/linodefirewall_types.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+	// +optional
+	// +kubebuilder:default=false
+	Enabled bool `json:"enabled,omitempty"`
+
+	// +kubebuilder:validation:MinLength=3
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	// +optional
+	Label string `json:"label,omitempty"`
```
Comment (komer3) [inline api/v1alpha2/linodefirewall_types.go:45]: question: what do we use this label for? Custom naming for the firewall?

File: cloud/scope/firewall.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+	// Override the controller credentials with ones from the Firewall's Secret reference (if supplied).
+	if params.LinodeFirewall.Spec.CredentialsRef != nil {
+		// TODO: This key is hard-coded (for now) to match the externally-managed `manager-credentials` Secret.
+		apiToken, err := getCredentialDataFromRef(ctx, params.Client, *params.LinodeFirewall.Spec.CredentialsRef, params.LinodeFirewall.GetNamespace(), "apiToken")
+		if err != nil {
+			return nil, fmt.Errorf("credentials from secret ref: %w", err)
+		}
+		linodeClientConfig.Token = string(apiToken)
+	}
```
Comment (komer3) [inline cloud/scope/firewall.go:69]: Should we have an else statement here incase the credential ref is not set for whatever reason?

File: .golangci.yml (hunk 2 above)
Comment (AshleyDumaine) [reply .golangci.yml:134]: Yeah, we would have needed a lot of nodupl like we did for the PG controller. We're duplicating a lot of code between PG, FW, and VPC controllers. A separate PR would be needed for using generics

File: cloud/scope/firewall.go (hunk 4 above)
Comment (komer3) [inline cloud/scope/firewall.go:69]: Should we have an else statement here incase the credential ref is not set for whatever reason?

File: .github/workflows/e2e-test.yaml (hunk 1 above)
Comment (AshleyDumaine) [reply .github/workflows/e2e-test.yaml:32]: I only see flavor tests there, none of the flavors include a firewall yet

File: api/v1alpha2/linodefirewall_types.go (hunk 3 above)
Comment (AshleyDumaine) [reply api/v1alpha2/linodefirewall_types.go:45]: I think that's leftover from March and can probably be removed. IIRC we make the Linode label based on the k8s resource name

File: .github/workflows/e2e-test.yaml (hunk 1 above)
Comment (komer3) [reply .github/workflows/e2e-test.yaml:32]: oh yeah we don't have firewall flavor yet. I though you added it. Sorry about that. 

e2e test you added runs under quick test so all good!

File: config/crd/kustomization.yaml [hunk 5]
```diff
@@ -36,6 +37,7 @@ patches:
 - path: patches/webhook_in_linodeobjectstoragebuckets.yaml
 - path: patches/webhook_in_linodeclustertemplates.yaml
 - path: patches/webhook_in_linodemachinetemplates.yaml
+#- path: patches/webhook_in_linodefirewalls.yaml
```
Comment (komer3) [inline config/crd/kustomization.yaml:40]: I'm guessing webhooks are not part of this PR implemention?

File: controller/linodefirewall_controller.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+			string(failureReason),
+			err.Error(),
+			reconciler.DefaultTimeout(r.ReconcileTimeout, reconciler.DefaultFWControllerReconcileTimeout),
+		)
+		r.Recorder.Event(fwScope.LinodeFirewall, corev1.EventTypeWarning, string(failureReason), err.Error())
+
+		switch {
+		case errors.Is(err, errTooManyIPs):
+			// Cannot reconcile firewall with too many ips, wait for an update to the spec
+			return ctrl.Result{}, nil
```
Comment (komer3) [inline controller/linodefirewall_controller.go:180]: question: I'm curious, how many ips is too many?

File: cloud/scope/firewall.go (hunk 4 above)
Comment (AshleyDumaine) [reply cloud/scope/firewall.go:69]: the credential ref is optional, we shouldn't need to do that. This is what the VPC and PG scopes are also doing.

**Review by komer3 — COMMENTED** [review]
Not sure why the e2e test for firewall is failing. But otherwise looks good!

**Review by komer3 — DISMISSED** [review]

File: controller/linodefirewall_controller.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+		return fmt.Errorf("failed to create mapper for LinodeFirewalls: %w", err)
+	}
+	err = ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1alpha2.LinodeFirewall{}).
+		WithOptions(options).
+		WithEventFilter(
+			predicate.And(
+				predicates.ResourceNotPausedAndHasFilterLabel(mgr.GetLogger(), r.WatchFilterValue),
+				predicate.GenerationChangedPredicate{},
+				predicate.Funcs{DeleteFunc: func(e event.DeleteEvent) bool { return false }},
```
Comment (eljohnson92) [inline controller/linodefirewall_controller.go:249]: I think we've landed on this predicate no longer being necessary?

Comment (AshleyDumaine) [reply controller/linodefirewall_controller.go:249]: yep

**Review by eljohnson92 — APPROVED** [review]
